lib/IRGen/GenCall.cpp

```diff
@@ -1515,13 +1705,323 @@ void irgen::extractScalarResults(IRGenFunction &IGF, llvm::Type *bodyType,
+        fieldTy = layout.getLocalContextType();
+      } else {
+        fieldLayout = layout.getElement(index + baseParameterIndex);
+        convention = layout.getArgumentConvention(index);
+        fieldTy = layout.getArgumentType(index);
+      }
+      Address fieldAddr =
+          fieldLayout->project(IGF, context, /*offsets*/ llvm::None);
+      storeArgumentIntoContext(llArgs, fieldTy,
+                               /*fieldTI*/ fieldLayout->getType(), fieldAddr,
+                               *convention, IGF, /*needsLoad*/ true,
+                               isOutlined);
+    }
+    auto indirectResults =
+        getCallee().getSubstFunctionType()->getIndirectFormalResults();
+    unsigned index = 0;
+    for (auto indirectResult : indirectResults) {
+      (void)indirectResult;
+      auto &fieldLayout = layout.getElement(index + baseIndirectReturnIndex);
+      Address fieldAddr =
+          fieldLayout.project(IGF, context, /*offsets*/ llvm::None);
+      cast<LoadableTypeInfo>(fieldLayout.getType())
+        .initialize(IGF, llArgs, fieldAddr, isOutlined);
+      ++index;
+    }
+    if (layout.hasBindings()) {
+      auto bindingLayout = layout.getBindingsLayout();
+      auto bindingsAddr = bindingLayout.project(IGF, context, /*offsets*/ None);
+      layout.getBindings().save(IGF, bindingsAddr);
+    }
+    // At this point, llArgs contains the arguments that are being passed along
+    // via the async context.  We can safely drop them on the floor.
+    (void)llArgs.claimAll();
+    // TODO: Validation: we should be able to check that the contents of llArgs
+    //       matches what is expected by the layout.
+  }
+  Address getErrorSlot(SILType errorType) override {
+    auto layout = getAsyncContextLayout();
+    auto errorLayout = layout.getErrorLayout();
+    return errorLayout.project(IGF, context, /*offsets*/ llvm::None);
+  };
+};
+
+} // end anonymous namespace
+
+std::unique_ptr<CallEmission> irgen::getCallEmission(IRGenFunction &IGF,
+                                                     llvm::Value *selfValue,
+                                                     Callee &&callee) {
+  if (callee.getOrigFunctionType()->isAsync()) {
+    return std::make_unique<AsyncCallEmission>(IGF, selfValue,
+                                               std::move(callee));
+  } else {
+    return std::make_unique<SyncCallEmission>(IGF, selfValue,
+                                              std::move(callee));
+  }
+}
+
 /// Emit the unsubstituted result of this call into the given explosion.
 /// The unsubstituted result must be naturally returned directly.
 void CallEmission::emitToUnmappedExplosion(Explosion &out) {
+  assert(state == State::Emitting);
   assert(LastArgWritten == 0 && "emitting unnaturally to explosion");
 
   auto call = emitCallSite();
 
+  if (getCallee().getOrigFunctionType()->isAsync()) {
+    SILFunctionConventions fnConv(getCallee().getOrigFunctionType(),
```

Comment:
I think this should be the substituted function type.

Reply:
Fixed.